cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -185,25 +191,235 @@ fn preflight_host_function_or_maybe_panic(
+        &storage,
+        &budget,
+        &diagnostic_events,
+    )?;
+    let transaction_data_cstr = CString::new(transaction_data.to_xdr_base64()?)?;
     Ok(CPreflightResult {
         error: null_mut(),
         results: scvals_to_c(results)?,
-        footprint: fp_cstr.into_raw(),
+        transaction_data: transaction_data_cstr.into_raw(),
+        min_fee: min_fee,
         auth: recorded_auth_payloads_to_c(auth_payloads)?,
-        events: host_events_to_c(events)?,
+        events: diagnostic_events_to_c(diagnostic_events)?,
         cpu_instructions: budget.get_cpu_insns_count(),
         memory_bytes: budget.get_mem_bytes_count(),
     })
 }
 
+fn compute_transaction_data_and_min_fee(
+    invoke_hf_op: &InvokeHostFunctionOp,
+    snapshot_source: &CSnapshotSource,
+    storage: &Storage,
+    budget: &Budget,
+    events: &Vec<DiagnosticEvent>,
+) -> Result<(SorobanTransactionData, i64), Box<dyn error::Error>> {
+    let soroban_resources = calculate_soroban_resources(snapshot_source, storage, budget, events)?;
+    let fee_configuration = get_fee_configuration(snapshot_source)?;
+
+    let transaction_resources = TransactionResources {
+        instructions: soroban_resources.instructions,
+        read_entries: soroban_resources.footprint.read_only.as_vec().len() as u32,
+        write_entries: soroban_resources.footprint.read_write.as_vec().len() as u32,
+        read_bytes: soroban_resources.read_bytes,
+        write_bytes: soroban_resources.write_bytes,
+        metadata_size_bytes: soroban_resources.extended_meta_data_size_bytes,
+        // Note: we could get a better transaction size if the full transaction was passed down to libpreflight
+        transaction_size_bytes: estimate_max_transaction_size(
+            invoke_hf_op,
+            &soroban_resources.footprint,
+        )?,
+    };
+    let (min_fee, ref_fee) =
+        compute_transaction_resource_fee(&transaction_resources, &fee_configuration);
+    let transaction_data = SorobanTransactionData {
+        resources: soroban_resources,
+        refundable_fee: ref_fee,
+        ext: ExtensionPoint::V0,
+    };
+    Ok((transaction_data, min_fee))
+}
+
+fn estimate_max_transaction_size(
+    invoke_hf_op: &InvokeHostFunctionOp,
+    fp: &LedgerFootprint,
+) -> Result<u32, Box<dyn error::Error>> {
+    let source = MuxedAccount::MuxedEd25519(MuxedAccountMed25519 {
+        id: 0,
+        ed25519: Uint256([0; 32]),
+    });
+    // generate the maximum memo size and signature size
+    // TODO: is this being too conservative?
+    let memo_text: Vec<u8> = [0; 28].into();
+    // TODO: find a better way to do this:
+    let mut signatures: Vec<DecoratedSignature> = vec![];
+    let mut signatures_left = 20;
+    while signatures_left > 0 {
+        signatures.push(DecoratedSignature {
+            hint: SignatureHint([0; 4]),
+            signature: Default::default(),
+        });
+        signatures_left -= 1;
+    }
+    let envelope = TransactionV1Envelope {
```

Comment:
I'm actually looking only at the internal `Transaction` size, specifically to avoid the signatures coupling (not sure if that's 100% correct, but that's how it behaves now).